docs/design.md

```diff
@@ -0,0 +1,106 @@
+Confidential Containers. More specifically, Confidential Containers requires:
+
+- Container image is always protected, e.g, signed and/or encrypted.
+- Download and unpack container image to the secure and attested enclave for the actual workloads
+to run with the protections of confidentiality and integrity.
+- Prove that sensitive workloads are running on a genuine and trusted hardware TEE.
+- Follow up trust and threat model defined by Confidential Containers.
+
+Enclave-CC approach would strictly align with above design principles to implement a general
+process-based Confidential Containers co-existing with VM-based Confidential Containers. It
+provides a consistent Kubernetes user experience across hardware TEEs, but also compared with the
+existing, non confidential Kubernetes usage patterns. The container workload is unmodified (but
+encrypted/signed) and identical across hardware TEEs.
+
+![k8s_ux.png](docs/images/k8s_ux.png)
+
+where:
+- [Inclavare Containers Project](https://github.com/inclavare-containers/inclavare-containers) is
+a container runtime to launch the process-based Confidential Container for Intel SGX.
+
+- LibOS generally refers to the SGX LibOS implementations, such as
+[Gramine](https://github.com/gramineproject/gramine) and
+[Occlum](https://github.com/occlum/occlum).
+
+- `rune` is the container runtime for spawning and running enclaves in containers.
+
+- `runelet` is the containerized process used to interact with LibOS through common PAL APIs.
+
+## Architecture
+
+![arch.png](docs/images/arch.png)
+
+enclave-CC introduces a helper enclave called `agent enclave` which is a fixed function enclave.
+It is launched by `shim-rune` during pod creation in a private method. More specifically,
+`agent enclave` is deployed in the form of OCI bundle rather than a container image, and thus
+`shim-rune` can call `rune` to start it. `agent enclave` has the same life cycle as the pod.
+Its main job is to receive and then load the actual workload's container image into an
+`app enclave`.
+
+To do this securely both the agent and app enclaves establish trust through attestation.
+
+`agent enclave` is responsible for hosting a LibOS loaded from OCI bundle previously mentioned
+and `enclave-agent` on top of the LibOS. `enclave-agent` consists of
+[image management service](https://github.com/confidential-containers/image-rs) and
+[attestation agent](https://github.com/confidential-containers/attestation-agent). They are the
+common pieces from Confidential Containers project.
+
+FUSE encryption filesystem acts as a temporary storage, providing the protections for the metadata
+and OCI bundle. There are many [alternatives](https://nuetzlich.net/gocryptfs/comparison/) to use.
+
+The LibOS running inside `app enclave` needs to establish an attested channel to retrieve the
+decryption key to mount FUSE encryption filesystem. The attested channel can be established in
+several ways according to policy:
```

Comment:
This needs to be discussed. The policy is global and should be provisioned in an early stage. This policy file is similar to kata-agent config file in kata-cc. However, it is still unclear how kata-cc deploys its config file.

Reply:
supporting different options sound fine, but it will add another layer of overhead/latency if the app enclave needs to verified remotely. sgx supports local attestation. the policy and manifest (in case local attestation) need to be build into agent enclave bundle to be verified by the remote KBS/verifier.

Reply:
Let's take the manifest for Gramine as example. It is per-enclave, so it would be provisioned remotely for app enclave. More specifically, the provisioning should happen, for example, along with retrieving the FUSE key through local attestation. So the problem is how the manifest for app enclave is transferred to agent enclave in advance.

In addition, the so-called policy is interpreted by LibOS. Currently there is no common definition for policy or manifest or whatever denoting a config among different LibOSes.